### PR TITLE
Make Key lower case by using HTTPHeaders

### PR DESCRIPTION
**Pull request by zirkome**
This changes the key mapper from TextMap to HTTPHeaders.

I integrated the work of @bajb in the commits to give him credit of the finding. 

I made this PR so it can be merged quicker as I need this fix in order to make it work with Zipkin.

### CODE REVIEW

**Comment by shouichi** [comment]
@bensigelman Could you take a look at the PR? Without this change, we'll get `rpc error: code = 13 desc = stream terminated by RST_STREAM with error code: 1` (I'm using zipkin).

File: go/otgrpc/shared.go [hunk 1]
```diff
@@ -18,6 +20,7 @@ type metadataReaderWriter struct {
 }
 
 func (w metadataReaderWriter) Set(key, val string) {
+	key = strings.ToLower(key)
```
Comment (bhs) [inline go/otgrpc/shared.go:23]: is this necessary? the HTTPHeaders carriers are supposed to be case-insensitive...

**Review by bhs — COMMENTED** [review]
LG aside from my question.

Thanks for `@`-mentioning me, I hadn't noticed this!

File: go/otgrpc/shared.go (hunk 1 above)
Comment (shouichi) [reply go/otgrpc/shared.go:23]: Tried removing this line and get an error `rpc error: code = 13 desc = stream terminated by RST_STREAM with error code: 1`.

Comment (shouichi) [inline go/otgrpc/shared.go]: Looks like opentracing-go implementation is case sensitive. I'll open a PR.

https://github.com/opentracing/opentracing-go/blob/master/propagation.go#L160

```
func (c HTTPHeadersCarrier) Set(key, val string) {
	h := http.Header(c)
	h.Add(key, val)
}
```

Comment (shouichi) [reply go/otgrpc/shared.go:23]: This line is needed. The following is the (summrised) zipkin-go-opentracing implementation.

```
func (t *tracerImpl) Inject(sc opentracing.SpanContext, format interface{}, carrier interface{}) error {
        switch format {
        case opentracing.TextMap, opentracing.HTTPHeaders:
                carrier.Set(zipkinTraceID, sc.TraceID.ToHex())
        }
}
```

It calls `Set` method regardless of format (TextMap or HTTPHeaders). So it seems grpc-opentracing is responsible to make keys case insensitive.

Comment (zirkome) [reply go/otgrpc/shared.go:23]: @bensigelman I agree with @shouichi, the java code is doing the same thing I think.

Comment (bhs) [reply go/otgrpc/shared.go:23]: @shouichi 
1. I now agree that this line is probably a good approach, but that's because GRPC (or, really, the HPACK impl) _is_ case-sensitive and requires lowercase.
2. Along those lines, the opentracing-go repo should not be changed IMO. For the true HTTP use case there is no need to adjust casing one way or the other.
3. There is "another way", which would be to use a _single_ key in the GRPC metadata and explicitly pack/unpack the contents of the TextMap into a string. That's not "hard", but is annoying as it would have to be done in every GRPC-OT language (and done with perfect consistency, including all annoying escaping and so forth).

TL;DR let's keep this code just as you have it, but let's _also_ add a comment explaining that the GRPC HPACK impl requires lowercase keys.

Thanks!

Comment (bhs) [reply go/otgrpc/shared.go:23]: Actually, I'm just going to merge this and I'll add the comment myself. Thanks again for the contribution!

Comment (shouichi) [reply go/otgrpc/shared.go:23]: Thank you! I think you can close #5 which tries to solve the same problem of this PR.
